source/language/openpulse.rst

```diff
@@ -0,0 +1,650 @@
+   getframe(str name) $q-> frame;
+
+For example, we can parmaeterize frames on different qubits using the wildcard reference
+.. code-block:: javascript
+
+   calconfig {
+     setframe("frame",  Frame(5.120e9, 0)) $0;
+     setframe("frame",  Frame(4.938e9, 0)) $1;
+    }
+
+   defcal access_frames $q {
+     frame carrier = getframe("frame") $q;
+   }
+
+Multiple frames associated with a qubit are uniquely accessed by their string identifier e.g.
+
+.. code-block:: javascript
+
+   calconfig {
+     setframe("frame_01",  Frame(5.0e9, 0)) $0;
+     setframe("frame_12",  Frame(6.0e9, 0)) $1;
+     ...
+    }
+
+   defcal access_frames $q {
+     qframe_01 = getframe("frame_01") $q;
+     qframe_12 = getframe("frame_12") $q;
+   }
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase) tuple of a frame can be manipulated throughout program
+by referencing ``.frequency``, and ``.phase``. Operations must be
+appropriate for the respective type, ``float`` for frequency and ``angle`` for
+phase. Again, the exact precision of these calculations is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
+``driveframe``:
+
+.. code-block:: javascript
+
+   // Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   driveframe.phase += pi/4;
+
+   // Define a calibration for the rz gate on all physical qubits
+
+   calconfig {
+     setframe("frame",  Frame(5.0e9, 0)) $0;
+     setframe("frame",  Frame(6.0e9, 0)) $1;
+     ...
+    }
+
+   defcal rz(angle[20] theta) $q {
+     frame rz_frame = getframe("frame") $q;
+     rz_frame.phase -= theta;
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: javascript
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Waveforms
+---------
+
+Waveforms are of type ``waveform`` and can either be:
+
+- An array of complex samples which define the points for the waveform envelope
+- An abstract mathematical function representing a waveform. This will later be
+  materialized into a list of complex samples, either by the compiler or the hardware
+  using the parameters provided to the pulse template.
+
+A value of type ``waveform`` is retrieved by explicitly constructing the complex samples
+or by calling one of the built-in waveform template functions. The latter are initialized by
+assigning a ``waveform`` to the result of a (kernel) function call. Note that each of these
+kernel functions takes a type ``length`` as a first argument, since waveforms need to have a
+definite length. Using the hardware dependent ``dt`` unit is recommended for this length,
+since the compiler may need to down-sample a higher precision waveform to physically realize it.
+
+.. code-block:: javascript
+
+   // arbitrary complex samples
+   waveform arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
```

Comment:
Note: Will need to define array and complex types either in OpenQASM or specifically at the OpenPulse level. There have been discussion for both at the OpenQASM level

Reply:
I would support adding both at the OpenQASM level.